torch/_inductor/codegen/cpp.py

```diff
@@ -357,6 +359,198 @@ def stride_at_vec_range(index: sympy.Expr, var: sympy.Symbol, vec_length: int):
+            _right_loop_nest = _cpp_kernel_proxy_list[idx + 1].loop_nest
+            if _get_parallel_depth(_cpp_kernel_proxy_list[idx]) != _get_parallel_depth(
+                _cpp_kernel_proxy_list[idx + 1]
+            ):
+                # Ensure same parallel depth.
+                # Refer to poolformer_m36 in timm_models:
+                # Node1 has loop collapsed(2), Node2 can't be loop collapsed due to it's reduction.
+                return False
+            if len(_left_loop_nest.root) != len(_right_loop_nest.root) or any(  # type: ignore[union-attr]
+                not _inner(left_root, right_root, _outer_loop_fusion_depth)
+                for left_root, right_root in zip(
+                    _left_loop_nest.root, _right_loop_nest.root  # type: ignore[union-attr]
+                )
+            ):
+                return False
+
+        return True
+
+    def _merge_outer_fusion_kernels(
+        self,
+        cpp_kernel_proxy_list,
+    ):
+        loop_nest_list: List[LoopNestWithSplit] = [
+            kernel.loop_nest for kernel in cpp_kernel_proxy_list
+        ]
+        metrics.cpp_outer_loop_fused_inner_counts.append(len(loop_nest_list))
+
+        def _merge_outer_fusion_loop_levels(
+            _loop_level_nested_list: List[List["LoopLevel"]],
+            _outer_loop_fusion_depth,
+        ):
+            assert _outer_loop_fusion_depth >= 1
+            if (_outer_loop_fusion_depth := _outer_loop_fusion_depth - 1) >= 1:
+                # Further merge the next loop level
+                # In certain outer loop fusion depth, still may have the
+                # splits of main loop and tile loop (refer to test_outer_loop_fusion_log_softmax in test_cpu_repro.py).
+                # We assume for each kernel, same loop split at certain outer loop fusion depth.
+                _loop_level_split_num = len(_loop_level_nested_list[0])
+                for _loop_level_split_idx in range(_loop_level_split_num):
+                    # Deal main loop and tail loop seperately
+                    _next_loop_level_nested_list = [
+                        _loop_level_list[_loop_level_split_idx].inner
+                        for _loop_level_list in _loop_level_nested_list
+                    ]
+                    _merge_outer_fusion_loop_levels(
+                        _next_loop_level_nested_list,
+                        _outer_loop_fusion_depth,
+                    )
+            else:
+                # This is the last looplevel to do outer loop fusion
+                _loop_level_split_num = len(_loop_level_nested_list[0])
+                for _loop_level_split_idx in range(_loop_level_split_num):
+                    # If at any outer loop fusion depth, we have split loop into main loop and tail loop,
+                    # The main loop and tail loop should have different OuterLoopFusedKernel.
+                    outer_loop_fused_kernel = OuterLoopFusedKernel()
```

Comment:
Do I understand correctly that you are constructing the kernel from:
```
main_loop1:
    main_kernel1
tail_loop1:
    tail_kernel1
main_loop2:
    main_kernel2
tail_loop2:
    tail_kernel2
```
to
```
outer_fused_kernel1:
    main_loop1:
        main_kernel1
    main_loop2:
        main_kernel2
outer_fused_kernel2:
    tail_loop1:
        tail_kernel1
    tail_loop2:
        tail_kernel2
```
But shouldn't it be a single outer fused kernel:
```
outer_fused_kernel:
    main_loop1:
        main_kernel1
    tail_loop1:
        tail_kernel1
    main_loop2:
        main_kernel2
    tail_loop2:
        tail_kernel2
```

Reply:
As we discussed offline, here is used to support the case when loop split happens at any outer loop fusion depth. We have simplified this code by blocking the loop split at outer loop fusion depth in `check_outer_fusion_loop_level_attr`.